sensor/common/networkflow/manager/manager_impl.go

```diff
@@ -1094,49 +1094,64 @@ func getIPAndPort(address *sensor.NetworkAddress) net.NetworkPeerID {
 	return tuple
 }
 
+func getConnection(conn *sensor.NetworkConnection) *connection {
+	var incoming bool
+	switch conn.Role {
+	case sensor.ClientServerRole_ROLE_SERVER:
+		incoming = true
+	case sensor.ClientServerRole_ROLE_CLIENT:
+		incoming = false
+	default:
+		return nil
+	}
+
+	remote := net.NumericEndpoint{
+		IPAndPort: getIPAndPort(conn.GetRemoteAddress()),
+		L4Proto:   net.L4ProtoFromProtobuf(conn.GetProtocol()),
+	}
+	local := getIPAndPort(conn.GetLocalAddress())
+
+	// Special handling for UDP ports - role reported by collector may be unreliable, so look at which port is more
+	// likely to be ephemeral. In case a port is set to 0, collector couldn't retrieve this value, we assume the
+	// connection works in the direction opposite of this port.
+	if remote.L4Proto == net.UDP {
+		incoming = netutil.IsEphemeralPort(remote.IPAndPort.Port) > netutil.IsEphemeralPort(local.Port)
+	}
+
+	return &connection{
+		local:       local,
+		remote:      remote,
+		containerID: conn.GetContainerId(),
+		incoming:    incoming,
+	}
+}
+
 func getUpdatedConnections(hostname string, networkInfo *sensor.NetworkConnectionInfo) map[connection]timestamp.MicroTS {
 	updatedConnections := make(map[connection]timestamp.MicroTS)
 
 	flowMetrics.NetworkFlowMessagesPerNode.With(prometheus.Labels{"Hostname": hostname}).Inc()
 
 	for _, conn := range networkInfo.GetUpdatedConnections() {
-		var incoming bool
-		switch conn.Role {
-		case sensor.ClientServerRole_ROLE_SERVER:
-			flowMetrics.NetworkFlowsPerNodeByType.With(prometheus.Labels{"Hostname": hostname, "Type": "incoming", "Protocol": conn.Protocol.String()}).Inc()
-			incoming = true
-		case sensor.ClientServerRole_ROLE_CLIENT:
-			flowMetrics.NetworkFlowsPerNodeByType.With(prometheus.Labels{"Hostname": hostname, "Type": "outgoing", "Protocol": conn.Protocol.String()}).Inc()
-			incoming = false
-		default:
+		c := getConnection(conn)
+		if c == nil {
```

Comment:
I think this might be more common than we might expect it to be, collector shouldn't be sending connections that are not marked as either server or client but I was seeing some unknowns when I was doing some testing. I'll check that it doesn't swamp the logs with garbage and implement one of the proposals if that is the case.

Reply:
Ended up adding an extra error return value and logging it in the caller with `Warn` severity, didn't see it triggering after some quick testing, so we are probably fine. Happy to change the severity if that one is not the correct one.